crates/amalthea/src/socket/shell.rs

```diff
@@ -145,7 +147,7 @@ impl Shell {
                 self.handle_request(req, |h, r| self.handle_comm_info_request(h, r))
             },
             Message::CommOpen(req) => self.handle_comm_open(req),
-            Message::CommMsg(req) => self.handle_request(req, |h, r| self.handle_comm_msg(h, r)),
+            Message::CommMsg(req) => self.handle_comm_msg(req),
```

Comment:
There are 2 very important things I want to call out here:
- Prior to this change, we were sending idle/busy messages _twice_ for each comm msg! Once in `handle_request()` and once in `handle_comm_msg()`! (any way to add an integration test?)
- Are we 100% sure that `handle_comm_msg()` does not need the lock on the `shell_handler`? It previously got the lock but was not actively using it, but could it have been passively using it?

Reply:
yep I noticed the double status too. I'll look into a test.

I don't think we need a lock on the shell handler if we're not calling shell methods. The shell handler will not be called by anyone else until the request handler returns.

Ideally we'd set up things in such a way we don't need a shell handler lock at all. Well after taking a quick look it does seem the shell thread can fully own the handler.

Reply:
Looks like our existing test purposefully ignored status messages: https://github.com/posit-dev/ark/blob/ba2bf26debd055e84c1bb61171dd956be06959f0/crates/amalthea/tests/client.rs#L403-L429

That's likely because there the ordering of the Idle and comm reply messages is undetermined since the latter is emitted on a different thread and there is no waiting mechanism for comm message handling.

I've refactored the loop so that we do test for status messages without introducing undeterminism.

Reply:
> Since Shell is supposed to be locked while the kernel is busy anyway, I think the kernel should own the handling. So the message handling would run on the kernel/RMain thread (which would own the shell socket) rather than on the shell thread (which would be removed).

But not all Shell messages require the lock right? Like `handle_comm_close`. So we'd be slowing those down potentially?

Reply:
There would be no lock, just blocking handling, and `handle_comm_close` would be a very quick handler.